src/main/scala/org/broadinstitute/hail/utils/TextTableReader.scala

```diff
@@ -54,45 +59,77 @@ case class TextTableConfiguration(
 object TextTableReader {
 
   val booleanRegex = """^([Tt]rue)|([Ff]alse)|(TRUE)|(FALSE)$"""
-  val intRegex = """^-?\d+$"""
-  val doubleRegex = """^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$"""
   val variantRegex = """^.+:\d+:[ATGC]+:([ATGC]+|\*)(,([ATGC]+|\*))*$"""
   val locusRegex = """^.+:\d+$"""
-  val headToTake = 20
-
-  def guessType(values: Seq[String], missing: String): Option[Type] = {
-    require(values.nonEmpty)
-
-    val size = values.size
-
-    val booleanMatch = values.exists(value => value.matches(booleanRegex))
-    val variantMatch = values.exists(value => value.matches(variantRegex))
-    val locusMatch = values.exists(value => value.matches(locusRegex))
-    val doubleMatch = values.exists(value => value.matches(doubleRegex))
-    val intMatch = values.exists(value => value.matches(intRegex))
-
-    val defined = values.filter(_ != missing)
-
-    val allBoolean = defined.forall(_.matches(booleanRegex))
-    val allVariant = defined.forall(_.matches(variantRegex))
-    val allLocus = defined.forall(_.matches(locusRegex))
-    val allDouble = defined.forall(_.matches(doubleRegex))
-    val allInt = defined.forall(_.matches(intRegex))
-
-    if (values.forall(_ == missing))
-      None
-    else if (allBoolean && booleanMatch)
-      Some(TBoolean)
-    else if (allVariant && variantMatch)
-      Some(TVariant)
-    else if (allLocus && locusMatch)
-      Some(TLocus)
-    else if (allInt && intMatch)
-      Some(TInt)
-    else if (allDouble && doubleMatch)
-      Some(TDouble)
-    else
-      Some(TString)
+  val doubleRegex = """^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$"""
+  val intRegex = """^-?\d+$"""
+
+  def imputeTypes(values: RDD[WithContext[String]], header: Array[String],
+    delimiter: String, missing: String): Array[Option[Type]] = {
+    val nFields = header.length
+    val regexes = Array(booleanRegex, variantRegex, locusRegex, intRegex, doubleRegex).map(Pattern.compile)
+
+    val regexTypes: Array[Type] = Array(TBoolean, TVariant, TLocus, TInt, TDouble)
+
+    val nRegex = regexes.length
+    val missingIndex = 2 * nRegex
+
+    def zero() = {
+      val m = MultiArray2.fill[Boolean](nFields, missingIndex + 1)(true)
+      for (i <- 0 until nFields)
+        for (j <- 0 until nRegex)
+          m.update(i, j, false)
+      m
+    }
+
+    val imputation = values.treeAggregate(zero())({ case (ma, line) =>
+      line.foreach { l =>
+        val split = l.split(delimiter)
+        if (split.length != nFields)
+          fatal(s"expected $nFields fields, but found ${ split.length }")
+
+        var i = 0
+        while (i < nFields) {
+          val field = split(i)
+          if (field != missing) {
+            var j = 0
+            while (j < nRegex) {
+              val matches = regexes(j).matcher(field).matches()
+              ma.update(i, j, ma(i, j) || matches)
+              ma.update(i, j + nRegex, ma(i, j + nRegex) && (field == missing || matches))
```

Comment:
We also talked about using `nregex + 1` values per field rather than `nregex*2 + 1` because we only need "did everything match" and "was everything missing".